core/src/main/java/org/elasticsearch/index/query/FilteredQueryBuilder.java

```diff
@@ -31,36 +36,113 @@
-        this.queryBuilder = queryBuilder;
-        this.filterBuilder = filterBuilder;
+    public FilteredQueryBuilder(QueryBuilder queryBuilder, QueryBuilder filterBuilder) {
+        this.queryBuilder = (queryBuilder != null) ? queryBuilder : generateDefaultQuery();
+        this.filterBuilder = (filterBuilder != null) ? filterBuilder : EmptyQueryBuilder.PROTOTYPE;
+    }
+
+    /** Returns the query to apply the filter to. */
+    public QueryBuilder query() {
+        return queryBuilder;
+    }
+
+    /** Returns the filter to apply to the query results. */
+    public QueryBuilder filter() {
+        return filterBuilder;
     }
 
     @Override
-    protected void doXContent(XContentBuilder builder, Params params) throws IOException {
-        builder.startObject(NAME);
-        if (queryBuilder != null) {
-            builder.field("query");
-            queryBuilder.toXContent(builder, params);
+    protected boolean doEquals(FilteredQueryBuilder other) {
+        return Objects.equals(queryBuilder, other.queryBuilder) &&
+                Objects.equals(filterBuilder, other.filterBuilder);
+    }
+
+    @Override
+    public int doHashCode() {
+        return Objects.hash(queryBuilder, filterBuilder);
+    }
+
+    @Override
+    public Query doToQuery(QueryParseContext parseContext) throws QueryParsingException, IOException {
+        Query result = null;
+
+        Query query = queryBuilder.toQuery(parseContext);
+        Query filter = filterBuilder.toQuery(parseContext);
+
+        if (query == null) {
+            // Most likely this query was parsed from the JSON query DSL - it parsed to null so we ignore
+            // the whole filtered query as there is nothing to filter on. See FilteredQueryParser for an example.
+            return null;
         }
-        if (filterBuilder != null) {
-            builder.field("filter");
-            filterBuilder.toXContent(builder, params);
+
+        if (filter == null || Queries.isConstantMatchAllQuery(filter)) {
+            // no filter, or match all filter
+            result = query;
+        } else if (Queries.isConstantMatchAllQuery(query)) {
+            // if its a match_all query, use constant_score
+            result = new ConstantScoreQuery(filter);
+        } else {
+            // use a BooleanQuery
+            result = Queries.filtered(query, filter);
         }
+
+        return result;
+    }
+
+    @Override
+    public QueryValidationException validate() {
+        QueryValidationException validationException = null;
+        if (queryBuilder == null) {
+            validationException = addValidationError("Query in FilteredQueryBuilder cannot be null.", validationException);
+        }
+        if (filterBuilder == null) {
+            validationException = addValidationError("Filter in FilteredQueryBuilder cannot be null.", validationException);
+        }
```

Comment:
I was having the same question in another query I was looking at. I think we made sure in constructors that the two builders are never null, but I was also wondering if if doesn't hurt having extra checks, in case e.g. some later change makes it possible to sneak in null query builders somehow. On the other hand, test should cover this. Sorry, undecided here, that's just what I had in mind in similar situation.